packages/permission-controller/src/PermissionController.ts

```diff
@@ -1991,6 +2110,214 @@ export class PermissionController<
+
+    const [mergedCaveats, caveatDiffMap] = caveatPairs.reduce(
+      ([caveats, diffMap], [leftCaveat, rightCaveat]) => {
+        const [newCaveat, diff] = this.#mergeCaveat(leftCaveat, rightCaveat);
+
+        if (newCaveat !== undefined && diff !== undefined) {
+          caveats.push(newCaveat);
+          diffMap[newCaveat.type] = diff;
+        } else {
+          caveats.push(leftCaveat);
+        }
+
+        return [caveats, diffMap];
+      },
+      [[], {}] as [CaveatConstraint[], CaveatDiffMap<CaveatConstraint>],
+    );
+
+    const mergedRightUniqueCaveats = rightUniqueCaveats.map((caveat) => {
+      const [newCaveat, diff] = this.#mergeCaveat(undefined, caveat);
+
+      if (newCaveat !== undefined && diff !== undefined) {
+        caveatDiffMap[newCaveat.type] = diff;
+        return newCaveat;
+      }
+      throw new InvalidEmptyCaveatMergeError(caveat.type);
+    });
+
+    const allCaveats = [
+      ...mergedCaveats,
+      ...leftUniqueCaveats,
+      ...mergedRightUniqueCaveats,
+    ];
+
+    const newPermission = {
+      ...leftPermission,
+      ...rightPermission,
+      ...(allCaveats.length > 0
+        ? { caveats: allCaveats as NonEmptyArray<CaveatConstraint> }
+        : {}),
+    };
+
+    return [newPermission, caveatDiffMap];
+  }
+
+  /**
+   * Merges two caveats of the same type. The task of merging the values of the
+   * two caveats is delegated to the corresponding caveat type's merger implementation.
+   *
+   * @param leftCaveat - The left-hand caveat to merge.
+   * @param rightCaveat - The right-hand caveat to merge.
+   * @returns The merged caveat and the diff between the two caveats.
+   */
+  #mergeCaveat(
+    leftCaveat: CaveatConstraint | undefined,
+    rightCaveat: CaveatConstraint,
+  ): [CaveatConstraint, CaveatDiff<Json>] | [] {
+    /* istanbul ignore if: This should be impossible */
+    if (leftCaveat !== undefined && leftCaveat.type !== rightCaveat.type) {
+      throw new CaveatMergeTypeMismatchError(leftCaveat.type, rightCaveat.type);
+    }
+
+    const merger = this.#expectGetCaveatMerger(rightCaveat.type);
+    const [newValue, diff] = merger(leftCaveat?.value, rightCaveat.value);
+
+    return newValue !== undefined && diff !== undefined
+      ? [
+          {
+            type: rightCaveat.type,
+            value: newValue,
+          },
+          diff,
+        ]
+      : [];
+  }
+
+  /**
+   * Given two permission objects, computes 3 sets:
+   * - The set of caveat pairs that are common to both permissions.
+   * - The set of caveats that are unique to the existing permission.
+   * - The set of caveats that are unique to the requested permission.
+   *
+   * Assumes that the caveat arrays of both permissions are valid.
+   *
+   * @param leftPermission - The left-hand permission.
+   * @param rightPermission - The right-hand permission.
+   * @returns The sets of caveat pairs and unique caveats.
+   */
+  #collectUniqueAndPairedCaveats(
```

Comment:
Nit: This probably doesn't need to be a class function since it doesn't access state

Reply:
Good point: 8bd9b2f8